Update README for JSON version. (#18)
siarheishut/route_manager: README.md

## Route manager

Bus Route Manager allows you to record information about various buses with their routes and stops, and later receive
information about it.

### Input

Input is represented as a JSON map with the following fields:

| Field         | Type  | Optional | Description                           |  
|---------------|-------|----------|---------------------------------------|  
| base_requests | array | No       | Requests for initializing a database. |   
| stat_requests | array | No       | Requests to the database.             |           

#### base_requests:

Each base_request in the array is a JSON map of one of the following types:

1. Bus request.

| Field        | Type   | Optional | Description                                                               |  
|--------------|--------|----------|---------------------------------------------------------------------------|  
| type         | string | No       | Must be equal to "Bus".                                                   |  
| name         | string | No       | The name of the bus. Must be unique.                                      |  

> | stops        | array  | No       | The route of the bus represented as an array of strings. Can be expanded. | 

Author: Done.

| is_roundtrip | bool   | No       | Indicates a circular route.                                               |  

> If route is circular. then it expands to the following form.   

> > **{stop1} > {stop2} > ... > {stopN} > {stopN-1} > ... > {stop1}**

Author: Done.

> Adds a bus with its route to the database.

Author: Done.

2. Stop request.

| Field          | Type   | Optional | Description                                                                                     |  
|----------------|--------|----------|-------------------------------------------------------------------------------------------------|  
| type           | string | No       | Must be equal to "Stop".                                                                        |  
| name           | string | No       | The name of the stop. Must be unique.                                                           |  
| latitude       | float  | No       | The latitude of the stop.                                                                       | 
| longitude      | float  | No       | The longitude of the stop.                                                                      |  

> | road_distances | map    | No       | Pairs {`stopX`, `distX`}, where integer `distX` is the road_length from stop `name` to `stopX`. |  

Reviewer: `a road length`

Author: Done.

> > Distance from {stopX} to {stop} = distance from {stop} to {stopX}, if road_length from {stopX} to {stop} wasn’t set
> > explicitly.  
> > If neither {stopX}-{stopY} nor {stopY}-{stopX} road_lengths were provided, the geo_distance (computed by coordinates)
> > will be considered the distance between the stops.

Reviewer: ```suggestion
> The distance from `A` to `B` is considered to be equal to the distance from `B` to `A`, if road length
> from `A` to `B` is not provided explicitly. 
> 
> If neither `A`->`B` nor `B`->`A` road lengths are provided, the distance between the stops is computed by coordinates.
```

Author: Done.

> Adds a stop with its coordinates and road lengths to the database.

Author: Done.

#### stat_requests:

Each stat_request in the array is a JSON map of one of the following types:

1. Bus request.

| Field | Type   | Optional | Description                                         |  
|-------|--------|----------|-----------------------------------------------------|  
| id    | int    | No       | Request ID. Must be unique.                         |  
| type  | string | No       | Must be equal to "Bus".                             | 
| name  | string | No       | The name of the bus for which you want to get info. |

2. Stop request.

| Field | Type   | Optional | Description                                          |  
|-------|--------|----------|------------------------------------------------------|  
| id    | int    | No       | Request ID. Must be unique.                          |  

> | type  | string | No       | Must be equal to Stop.                               |  

Author: Done.

| name  | string | No       | The name of the stop for which you want to get info. |

### Output:

> The output is represented as a JSON array, where each element is a response to related stat_request. Response is

Author: Done.

represented as a JSON map of one of the following types:

1. Response to Bus Request.

- `If the bus was successfully found:`

| Field             | Type  | Description                                                                                                                                              |  
|-------------------|-------|----------------------------------------------------------------------------------------------------------------------------------------------------------|  

> | route_length      | float | The length of the bus route.                                                                                                                             |  

Author: Done.

> | request_id        | int   | Request ID for the response.                                                                                                                             |  

Reviewer: An ID of the corresponding request

Author: Done.

> | stop_count        | int   | The count of stops on the bus route.                                                                                                                     |  

Reviewer: `A number of`

Author: Done.

> | unique_stop_count | int   | The count of unique stops on the bus route.                                                                                                              |  

Reviewer: `A number of`

Author: Done.

| curvature         | float | `route_length` / `geo_route_length`. `geo_route_length` is an ideal route length: distance between any stopX and stopY is computed by their coordinates. |

- `If the bus wasn't found:`

| Field         | Type   | Description                                      |  
|---------------|--------|--------------------------------------------------| 

> | request_id    | int    | Request ID for the response.                     | 

Reviewer: Same as above

Author: Done.

> | error_message | string | The error message which is equal to "not found". | 

Reviewer: `An error message, which`

2. Response to Stop Request.

- `If the stop is in the database, and it’s included in at least one route:`

| Field      | Type  | Description                                                         |  
|------------|-------|---------------------------------------------------------------------|  

> | buses      | array | All buses that include stop `name` listed in lexicographical order. |  

Reviewer: `All the buses that stop at the provided stop, listed in lexicographical order.`

Author: Done.

> | request_id | int   | Request ID for the response.                                        |

Reviewer: Same as above

Author: Done.

- `If the stop is in the database, but it’s not included in any route:`

| Field      | Type  | Description                  |  
|------------|-------|------------------------------|  
| buses      | array | An empty array.              |  

> | request_id | int   | Request ID for the response. |  

Reviewer: Combine this one with the one above. Name it `If the stop is present in the database`

Author: Done.

- `If the stop is not in the database:`

| Field         | Type   | Description                                      |  
|---------------|--------|--------------------------------------------------|  

> | request_id    | int    | Request ID for the response.                     |  

Reviewer: Same as above

Author: Done.

> | error_message | string | The error message which is equal to "not found". |  

Reviewer: `An error message, which`

Author: Done.

|--------------|--------|----------|----------------------------------------------------------|  
| type         | string | No       | Must be equal to "Bus".                                  |  
| name         | string | No       | The name of the bus. Must be unique.                     |  
| stops        | array  | No       | The route of the bus represented as an array of strings. | 
| is_roundtrip | bool   | No       | Indicates a circular route.                              |  

> if `is_roundtrip` is set to `true`, than `"stops" : ["stop1", "stop2", "stop3", "stop1"]` describes the route, that
> starts at `stop1`, then goes to `stop2`, `stop3`, and then returns back directly to `stop1`. Then this route
> repeats: `stop1 -> stop2 -> stop3 -> stop1`.
>
> if `is_roundtrip` is set to `false`, then `"stops" : ["stop1", "stop2", "stop3"]` describes the route, that starts
> at `stop1`, then goes to `stop2`, `stop3`, then back to `stop2` and finally returns to `stop1`. Then this route
> repeats: `stop1 -> stop2 -> stop3 -> stop2 -> stop1`.
>
> Note: if `is_roundtrip` is set to `true`, the first stop must be the last stop.

2. Stop request.

> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Adds a stop with its coordinates and road lengths to the database.

Author: Fixed.

#### stat_requests:

Each stat_request in the array is a JSON map of one of the following types:

1. Bus request.

| Field | Type   | Optional | Description                                         |  
|-------|--------|----------|-----------------------------------------------------|  
| id    | int    | No       | Request ID. Must be unique.                         |  
| type  | string | No       | Must be equal to "Bus".                             | 
| name  | string | No       | The name of the bus for which you want to get info. |

2. Stop request.

| Field | Type   | Optional | Description                                          |  
|-------|--------|----------|------------------------------------------------------|  
| id    | int    | No       | Request ID. Must be unique.                          |  
| type  | string | No       | Must be equal to "Stop".                             |  
| name  | string | No       | The name of the stop for which you want to get info. |

### Output:

The output is represented as a JSON array, where each element is a response to a related stat_request. Each response is

> is a JSON map of one of the following types:

Reviewer: double `is`

Author: Deleted.

> 
> #### stat_requests:
> 
> Each stat_request in the array is a JSON map of one of the following types:
> 
> 1. Bus request.
> 
> | Field | Type   | Optional | Description                                         |  
> |-------|--------|----------|-----------------------------------------------------|  
> | id    | int    | No       | Request ID. Must be unique.                         |  
> | type  | string | No       | Must be equal to "Bus".                             | 
> | name  | string | No       | The name of the bus for which you want to get info. |
> 
> 2. Stop request.
> 
> | Field | Type   | Optional | Description                                          |  
> |-------|--------|----------|------------------------------------------------------|  
> | id    | int    | No       | Request ID. Must be unique.                          |  
> | type  | string | No       | Must be equal to "Stop".                             |  
> | name  | string | No       | The name of the stop for which you want to get info. |
> 
> ### Output:
> 
> The output is represented as a JSON array, where each element is a response to a related stat_request. Each response is
> is a JSON map of one of the following types:
> 
> 1. Response to Bus Request.
> 
> - `If the bus was successfully found:`
> 
> | Field             | Type  | Description                                                                                                                                              |  
> |-------------------|-------|----------------------------------------------------------------------------------------------------------------------------------------------------------|  
> | route_length      | float | The length of the bus route in meters.                                                                                                                   |  
> | request_id        | int   | An ID of the corresponding request.                                                                                                                      |  
> | stop_count        | int   | A number of stops on the bus route.                                                                                                                      |  
> | unique_stop_count | int   | A number of unique stops on the bus route.                                                                                                               |  
> | curvature         | float | `route_length` / `geo_route_length`. `geo_route_length` is an ideal route length: distance between any stopX and stopY is computed by their coordinates. |
> 
> - `If the bus wasn't found:`
> 
> | Field         | Type   | Description                                      |  
> |---------------|--------|--------------------------------------------------| 
> | request_id    | int    | An ID of the corresponding request.              | 
> | error_message | string | An error message, which is equal to "not found". | 
> 
> 2. Response to Stop Request.
> 
> - `If the stop is in the database, and it’s included in at least one route:`
> 
> | Field      | Type  | Description                                                                    |  
> |------------|-------|--------------------------------------------------------------------------------|  
> | buses      | array | All the buses that stop at the provided stop, listed in lexicographical order. |  
> | request_id | int   | An ID of the corresponding request.                                            |
> 
> - `If the stop is present in the database:`
> 
> | Field      | Type  | Description                         |  
> |------------|-------|-------------------------------------|  
> | buses      | array | An empty array.                     |  
> | request_id | int   | An ID of the corresponding request. |  

Reviewer: Please, combine these two:

```suggestion
- `If the stop is present in the database:`

| Field      | Type  | Description                                                                    |  
|------------|-------|--------------------------------------------------------------------------------|  
| buses      | array | All the buses that stop at the provided stop, listed in lexicographical order. |  
| request_id | int   | An ID of the corresponding request.                                            |
```

Author: Done.
